gcloud/storage/api.py

```diff
@@ -184,6 +170,11 @@ class _BucketIterator(Iterator):
     """
 
     def __init__(self, connection, extra_params=None):
+        # If an implicit connection was intended, we pass along `None` to the
+        # Bucket() constructor as well.
```

Comment:
The iterator would use the bound connection and then bind it to every `Bucket()` created.

Then in a `Batch()`, those buckets would not use the batch, since they would have their own explicitly-bound connection.